jax_distribution/systems/ff_ippo_rware.py

```diff
@@ -0,0 +1,407 @@
+import flax.linen as nn
+import jax
+import jax.numpy as jnp
+import jumanji
+import numpy as np
+import optax
+from flax.core.frozen_dict import FrozenDict
+from flax.linen.initializers import constant, orthogonal
+from jumanji.env import Environment, State
+from jumanji.types import Observation, TimeStep
+from jumanji.wrappers import AutoResetWrapper
+from optax._src.base import OptState
+
+from jax_distribution.types import Transition
+from jax_distribution.utils.timing_utils import TimeIt
+from jax_distribution.wrappers.jumanji import LogWrapper, RwareMultiAgentWrapper
+
+
+class ActorCritic(nn.Module):
+    """Actor Critic Network."""
+
+    action_dim: Sequence[int]
+    activation: str = "tanh"
+
+    @nn.compact
+    def __call__(
+        self, observation: Observation
+    ) -> Tuple[distrax.Categorical, chex.Array]:
+        """Forward pass."""
+        x = observation.agents_view
+
+        if self.activation == "relu":
+            activation = nn.relu
+        else:
+            activation = nn.tanh
+
+        actor_output = nn.Dense(
+            64, kernel_init=orthogonal(np.sqrt(2)), bias_init=constant(0.0)
+        )(x)
+        actor_output = activation(actor_output)
+        actor_output = nn.Dense(
+            64, kernel_init=orthogonal(np.sqrt(2)), bias_init=constant(0.0)
+        )(actor_output)
+        actor_output = activation(actor_output)
+        actor_output = nn.Dense(
+            self.action_dim, kernel_init=orthogonal(0.01), bias_init=constant(0.0)
+        )(actor_output)
+
+        masked_logits = jnp.where(
+            observation.action_mask,
+            actor_output,
+            jnp.finfo(jnp.float32).min,
+        )
+        actor_policy = distrax.Categorical(logits=masked_logits)
+
+        critic_output = nn.Dense(
+            64, kernel_init=orthogonal(np.sqrt(2)), bias_init=constant(0.0)
+        )(x)
+        critic_output = activation(critic_output)
+        critic_output = nn.Dense(
+            64, kernel_init=orthogonal(np.sqrt(2)), bias_init=constant(0.0)
+        )(critic_output)
+        critic_output = activation(critic_output)
+        critic_output = nn.Dense(
+            1, kernel_init=orthogonal(1.0), bias_init=constant(0.0)
+        )(critic_output)
+
+        return actor_policy, jnp.squeeze(critic_output, axis=-1)
+
+
+def get_learner_fn(
+    env: jumanji.Environment, apply_fn: Callable, update_fn: Callable, config: dict
+) -> Callable:
+    """Get the learner function."""
+
+    def _update_step(runner_state: Tuple, unused_target: Any) -> Tuple:
+        """Update the network."""
+
+        def _env_step(runner_state: Tuple, unused: Any) -> Tuple:
+            """Step the environment."""
+            params, opt_state, rng, env_state, last_timestep = runner_state
+
+            # SELECT ACTION
+            rng, _rng = jax.random.split(rng)
+            actor_policy, value = apply_fn(params, last_timestep.observation)
+            action = actor_policy.sample(seed=_rng)
```

Comment:
Name the keys so we know what they're used for

```suggestion
            rng, policy_rng = jax.random.split(rng)
            actor_policy, value = apply_fn(params, last_timestep.observation)
            action = actor_policy.sample(seed=policy_rng)
```